docs/developer/designs/topology-awareness/README.md

```diff
@@ -0,0 +1,306 @@
+type TopologyDomainInfo struct {
+    // Unique ID of this domain
+    ID TopologyDomainID
+
+    // Name of this domain
+    Name string
+
+    // Level in the hierarchy (e.g., "datacenter", "zone", "rack", "node")
+    Level string
+
+    // Parent domain, nil for root
+    Parent *TopologyDomainInfo
+
+    // Child domains
+    Children []*TopologyDomainInfo
+
+    // Nodes that belong to this domain
+    Nodes map[string]*node_info.NodeInfo
+
+    // Total available resources in this domain
+    AvailableResources v1.ResourceList
+
+    // Total allocated resources in this domain
+    AllocatedResources v1.ResourceList
+
+    // Depth in the tree from root (0 for root)
+    Depth int
+}
+```
+
+There are two possibilities as to where the struct should be built and kept:
+  * Plugin - We will use a plugin that will build it in `OnSessionOpen` and will use hooks similar to DRA plugin to keep tabs on all the changes made in the cluster.
+  * ClusterInfo - This can be done with TopologyInfo structs in the ClusterInfo package, the plugin will later read this from the cache to make node order decision and will update this struct with simulated allocations.
+Because this might be resource intensive we will leave users with the option to totally disable it and only reference it in the plugin.
+
+#### Topology Awareness Plugin
+
+##### Setup
+Maintain topology tree with available resources using the Allocate/Deallocate event handlers (similar to the dynamicresources plugin)
+
+##### Pre Process
+Register the PrePredicate function (Possibly change its name to PreFilter ? ) to create a job-specific and look for allocation in that tree.
+
+####  Node Ordering and Filtering
+- This is currently called per task, so with the current setup it will have to cache the result in the plugin (Possibly also using the resources tree hash?)
+- Search for the best allocation of the job in the tree and save that too for the next Predicate and NodeOrder calls.
+    - Predicate function will only be used for strict topology-aware scheduling.
+- NodeOrder will rank the selected node above all and then other nodes by distance. Use a factor that will make this plugin more dominant the any other (similar to pod affinity node order).
+- For future extensions: New plugin point for `BeforeBind` will allow to mutate the bind request / pod creating the bind request. It will be used to add rank labels to the pod in order to specify its chosen rank inside the job.
+
+#### Pod Grouper Implementation
+
+The pod-grouper component will be enhanced to:
+
+1. Look for topology-related annotations on the top owner of pods
+2. Extract and validate topology constraint information
+3. Set the corresponding fields in the PodGroup that it creates
+4. Support both Required and Preferred constraints
+
+For pods that are part of the same job but have different topology requirements, the pod-grouper will use the most restrictive constraints to ensure all pods can be scheduled.
```

Comment:
Since we said that the topology information is extracted from the top owner of the pod, pods should probably not be part of it.